File: packages/sdk-ts/tests/response.test.ts
import { describe, expect, expectTypeOf, it } from "vitest";
import { z } from "zod/v4";
import type { RPCMethod } from "@browserbasehq/stagehand-protocol/json-rpc/schemas";
import { StagehandMethods } from "@browserbasehq/stagehand-protocol/schema-registry";
import type { NavigationResponseDescriptor } from "@browserbasehq/stagehand-protocol/types";
import { Page, Response } from "../src/index.js";
import { RPCClient } from "../src/rpcClient.js";

class FakeResponseClient extends RPCClient {
  readonly calls: Array<{ method: string; params: unknown }> = [];
  readonly responses = new Map<string, unknown[]>();
  readonly preserveQueuedResponseIdentity: boolean;

  constructor(options: { preserveQueuedResponseIdentity?: boolean } = {}) {
    super({
      serviceWorker: {
        targetId: "worker-target",
        url: "chrome-extension://stagehand/service-worker.js",
        title: "Stagehand",
        extensionId: "stagehand",
      },
      send: async () => {},
      close: () => {},
    });
    this.preserveQueuedResponseIdentity = options.preserveQueuedResponseIdentity ?? false;
  }

  queue<Method extends RPCMethod>(
    method: Method,
    response: z.input<Method["result"]> | Error,
  ): void {
    const responses = this.responses.get(method.name) ?? [];
    responses.push(response);
    this.responses.set(method.name, responses);
  }

  async send<Method extends RPCMethod>(
    method: Method,
    params: z.input<Method["params"]>,
  ): Promise<z.output<Method["result"]>> {
    this.calls.push({ method: method.name, params });
    const response = this.responses.get(method.name)?.shift();
    if (response === undefined) throw new Error(`No fake response queued for ${method.name}`);
    if (response instanceof Error) throw response;
    if (this.preserveQueuedResponseIdentity) {
      return response as z.output<Method["result"]>;
    }
    return method.result.parse(response) as z.output<Method["result"]>;
  }
}

const descriptor: NavigationResponseDescriptor = {
  responseId: "response-1",
  url: "https://example.test/final",
  status: 201,
  statusText: "Created",
  headers: { "content-type": "application/json" },
  fromServiceWorker: true,
};

describe("TypeScript Response", () => {
  it("copies its input descriptor", async () => {
    const client = new FakeResponseClient();
    const input: NavigationResponseDescriptor = {
      ...descriptor,
      headers: { ...descriptor.headers },
    };
    const response = new Response(client, input);
    client.queue(StagehandMethods.responseFinished, { error: null });

    input.responseId = "mutated-response";
    input.url = "https://mutated.test";
    input.status = 500;
    input.statusText = "Mutated";
    input.headers["content-type"] = "mutated";
    input.fromServiceWorker = false;

    expect(response.url()).toBe("https://example.test/final");
    expect(response.status()).toBe(201);
    expect(response.statusText()).toBe("Created");
    expect(response.headers()).toStrictEqual({ "content-type": "application/json" });
    expect(response.fromServiceWorker()).toBe(true);
    await expect(response.finished()).resolves.toBeNull();
    expect(client.calls).toStrictEqual([
      { method: "response.finished", params: { responseId: "response-1" } },
    ]);
  });

  it("exposes immediate metadata locally with defensive header copies", () => {
    const client = new FakeResponseClient();
    const response = new Response(client, descriptor);

    expect(response.url()).toBe("https://example.test/final");
    expect(response.status()).toBe(201);
    expect(response.statusText()).toBe("Created");
    expect(response.ok()).toBe(true);
    expect(response.fromServiceWorker()).toBe(true);

    const headers = response.headers();
    headers["content-type"] = "mutated";
    expect(response.headers()).toStrictEqual({ "content-type": "application/json" });
    expect(client.calls).toStrictEqual([]);
  });

  it("retrieves headers and connection metadata lazily", async () => {
    const client = new FakeResponseClient({ preserveQueuedResponseIdentity: true });
    const response = new Response(client, descriptor);
    const headerArray = {
      headers: [
        { name: "Set-Cookie", value: "first=1" },
        { name: "set-cookie", value: "second=2" },
      ],
    };
    client.queue(StagehandMethods.responseAllHeaders, {
      headers: { "Content-Type": "application/json", "Set-Cookie": "first=1\nsecond=2" },
    });
    client.queue(StagehandMethods.responseHeadersArray, headerArray);
    client.queue(StagehandMethods.responseHeadersArray, headerArray);
    client.queue(StagehandMethods.responseHeadersArray, headerArray);
    client.queue(StagehandMethods.responseSecurityDetails, {
      value: {
        issuer: "Example CA",
        protocol: "TLS 1.3",
        subjectName: "example.test",
        validFrom: 1,
        validTo: 2,
      },
    });
    client.queue(StagehandMethods.responseServerAddr, {
      value: { ipAddress: "203.0.113.10", port: 443 },
    });

    await expect(response.allHeaders()).resolves.toStrictEqual({
      "Content-Type": "application/json",
      "Set-Cookie": "first=1\nsecond=2",
    });
    const headers = await response.headersArray();
    headers[0]!.value = "mutated";
    expect(headerArray.headers[0]!.value).toBe("first=1");
    await expect(response.headerValue("SET-cookie")).resolves.toBe("first=1, second=2");
    await expect(response.headerValues("set-COOKIE")).resolves.toStrictEqual([
      "first=1",
      "second=2",
    ]);
    await expect(response.securityDetails()).resolves.toStrictEqual({
      issuer: "Example CA",
      protocol: "TLS 1.3",
      subjectName: "example.test",
      validFrom: 1,
      validTo: 2,
    });
    await expect(response.serverAddr()).resolves.toStrictEqual({
      ipAddress: "203.0.113.10",
      port: 443,
    });

    expect(client.calls.map((call) => call.method)).toStrictEqual([
      "response.all_headers",
      "response.headers_array",
      "response.headers_array",
      "response.headers_array",
      "response.security_details",
      "response.server_addr",
    ]);
    expect(
      client.calls.every(
        (call) => (call.params as { responseId?: string }).responseId === "response-1",
      ),
    ).toBe(true);
  });

  it("decodes body, text, and JSON through independent RPC calls", async () => {
    const client = new FakeResponseClient();
    const response = new Response(client, descriptor);
    for (let index = 0; index < 3; index += 1) {
      client.queue(StagehandMethods.responseBody, {
        body: "eyJvayI6dHJ1ZX0=",
        base64Encoded: true,
      });
    }

    await expect(response.body()).resolves.toStrictEqual(new TextEncoder().encode('{"ok":true}'));
    await expect(response.text()).resolves.toBe('{"ok":true}');
    await expect(response.json<{ ok: boolean }>()).resolves.toStrictEqual({ ok: true });
    expect(client.calls.map((call) => call.method)).toStrictEqual([
      "response.body",
      "response.body",
      "response.body",
    ]);
  });

  it("surfaces finished failures, transport errors, and malformed bodies", async () => {
    const client = new FakeResponseClient();
    const response = new Response(client, descriptor);
    client.queue(StagehandMethods.responseFinished, {
      error: { message: "net::ERR_FAILED" },
    });
    client.queue(StagehandMethods.responseBody, {
      body: "%%%",
      base64Encoded: true,
    });
    client.queue(StagehandMethods.responseBody, {
      body: "Zh==",
      base64Encoded: true,
    });
    client.queue(StagehandMethods.responseAllHeaders, new Error("handle unavailable"));

    const finishedError = await response.finished();
    expect(finishedError).toBeInstanceOf(Error);
    expect(finishedError?.message).toBe("net::ERR_FAILED");
    await expect(response.body()).rejects.toThrow("response.body returned invalid base64");
    await expect(response.body()).rejects.toThrow("response.body returned invalid base64");
    await expect(response.allHeaders()).rejects.toThrow("handle unavailable");
  });

  it("declares navigation methods as nullable response promises", () => {
    expectTypeOf<Page["goto"]>().returns.toEqualTypeOf<Promise<Response | null>>();
    expectTypeOf<Page["reload"]>().returns.toEqualTypeOf<Promise<Response | null>>();
    expectTypeOf<Page["goBack"]>().returns.toEqualTypeOf<Promise<Response | null>>();
    expectTypeOf<Page["goForward"]>().returns.toEqualTypeOf<Promise<Response | null>>();
  });
});

File: packages/sdk-go/response_test.go
package stagehand

import (
	"bytes"
	"context"
	"errors"
	"reflect"
	"strings"
	"testing"
)

func responseDescriptor() NavigationResponseDescriptor {
	return NavigationResponseDescriptor{
		ResponseID:        "response-1",
		URL:               "https://example.test/final",
		Status:            201,
		StatusText:        "Created",
		Headers:           map[string]string{"content-type": "application/json"},
		FromServiceWorker: true,
	}
}

func TestResponseExposesImmediateMetadataWithDefensiveHeaderCopies(t *testing.T) {
	t.Parallel()

	rpc := &recordingProtocolClient{}
	descriptor := responseDescriptor()
	response := newResponse(rpc, descriptor)
	descriptor.Headers["content-type"] = "mutated before access"

	if response.URL() != "https://example.test/final" {
		t.Fatalf("URL() = %q", response.URL())
	}
	if response.Status() != 201 || response.StatusText() != "Created" {
		t.Fatalf("status = (%d, %q)", response.Status(), response.StatusText())
	}
	if !response.OK() || !response.FromServiceWorker() {
		t.Fatal("response metadata convenience methods returned false")
	}

	headers := response.Headers()
	headers["content-type"] = "mutated after access"
	if got := response.Headers()["content-type"]; got != "application/json" {
		t.Fatalf("Headers()[content-type] = %q", got)
	}
	if len(rpc.calls) != 0 {
		t.Fatalf("immediate metadata made RPC calls: %#v", rpc.calls)
	}
}

func TestResponseRetrievesHeadersAndConnectionMetadataLazily(t *testing.T) {
	t.Parallel()

	rpc := &recordingProtocolClient{responses: map[string]any{
		"response.all_headers": ResponseAllHeadersResult{Headers: map[string]string{
			"Content-Type": "application/json",
			"Set-Cookie":   "first=1\nsecond=2",
		}},
		"response.headers_array": ResponseHeadersArrayResult{Headers: []NavigationHeader{
			{Name: "Set-Cookie", Value: "first=1"},
			{Name: "set-cookie", Value: "second=2"},
			{Name: "X-Empty", Value: ""},
		}},
		"response.security_details": ResponseSecurityDetailsResult{Value: &NavigationSecurityDetails{
			Issuer:      "Example CA",
			Protocol:    "TLS 1.3",
			SubjectName: "example.test",
			ValidFrom:   1,
			ValidTo:     2,
		}},
		"response.server_addr": ResponseServerAddrResult{Value: &NavigationServerAddr{
			IPAddress: "203.0.113.10",
			Port:      443,
		}},
	}}
	response := newResponse(rpc, responseDescriptor())
	ctx := context.Background()

	allHeaders, err := response.AllHeaders(ctx)
	if err != nil || allHeaders["Set-Cookie"] != "first=1\nsecond=2" {
		t.Fatalf("AllHeaders() = (%#v, %v)", allHeaders, err)
	}
	allHeaders["Set-Cookie"] = "mutated"

	headers, err := response.HeadersArray(ctx)
	if err != nil {
		t.Fatalf("HeadersArray() error = %v", err)
	}
	headers[0].Value = "mutated"
	value, present, err := response.HeaderValue(ctx, "SET-cookie")
	if err != nil || !present || value != "first=1, second=2" {
		t.Fatalf("HeaderValue() = (%q, %t, %v)", value, present, err)
	}
	values, err := response.HeaderValues(ctx, "set-COOKIE")
	if err != nil || !reflect.DeepEqual(values, []string{"first=1", "second=2"}) {
		t.Fatalf("HeaderValues() = (%#v, %v)", values, err)
	}
	empty, present, err := response.HeaderValue(ctx, "x-empty")
	if err != nil || !present || empty != "" {
		t.Fatalf("HeaderValue(x-empty) = (%q, %t, %v)", empty, present, err)
	}
	missing, present, err := response.HeaderValue(ctx, "missing")
	if err != nil || present || missing != "" {
		t.Fatalf("HeaderValue(missing) = (%q, %t, %v)", missing, present, err)
	}

	securityDetails, err := response.SecurityDetails(ctx)
	if err != nil || securityDetails == nil || securityDetails.SubjectName != "example.test" {
		t.Fatalf("SecurityDetails() = (%#v, %v)", securityDetails, err)
	}
	serverAddr, err := response.ServerAddr(ctx)
	if err != nil || serverAddr == nil || serverAddr.IPAddress != "203.0.113.10" {
		t.Fatalf("ServerAddr() = (%#v, %v)", serverAddr, err)
	}

	for _, call := range rpc.calls {
		if params, ok := call.params.(ResponseIDParams); !ok || params.ResponseID != "response-1" {
			t.Fatalf("RPC params = %#v", call.params)
		}
	}
}

func TestResponseBodyTextAndJSONUseIndependentRPCCalls(t *testing.T) {
	t.Parallel()

	rpc := &recordingProtocolClient{responses: map[string]any{
		"response.body": ResponseBodyResult{
			Body:          "eyJvayI6dHJ1ZX0=",
			Base64Encoded: true,
		},
	}}
	response := newResponse(rpc, responseDescriptor())
	ctx := context.Background()

	body, err := response.Body(ctx)
	if err != nil || !bytes.Equal(body, []byte(`{"ok":true}`)) {
		t.Fatalf("Body() = (%q, %v)", body, err)
	}
	text, err := response.Text(ctx)
	if err != nil || text != `{"ok":true}` {
		t.Fatalf("Text() = (%q, %v)", text, err)
	}
	var decoded struct {
		OK bool `json:"ok"`
	}
	if err := response.JSON(ctx, &decoded); err != nil || !decoded.OK {
		t.Fatalf("JSON() = (%#v, %v)", decoded, err)
	}
	if got := len(rpc.calls); got != 3 {
		t.Fatalf("response.body RPC calls = %d, want 3", got)
	}
}

func TestResponseSurfacesFinishedTransportAndMalformedBodyErrors(t *testing.T) {
	t.Parallel()

	rpc := &recordingProtocolClient{
		responses: map[string]any{
			"response.finished": ResponseFinishedResult{
				Error: &NavigationFinishedError{Message: "net::ERR_FAILED"},
			},
			"response.body": ResponseBodyResult{Body: "%%%", Base64Encoded: true},
		},
		callErrors: map[string]error{
			"response.all_headers": errors.New("handle unavailable"),
		},
	}
	response := newResponse(rpc, responseDescriptor())
	ctx := context.Background()

	if err := response.Finished(ctx); err == nil || err.Error() != "net::ERR_FAILED" {
		t.Fatalf("Finished() error = %v", err)
	}
	if _, err := response.Body(ctx); err == nil || !strings.Contains(err.Error(), "invalid base64") {
		t.Fatalf("Body() error = %v", err)
	}
	if _, err := response.AllHeaders(ctx); err == nil || err.Error() != "handle unavailable" {
		t.Fatalf("AllHeaders() error = %v", err)
	}
}

func TestResponseFinishedAndOptionalMetadataCanReturnNil(t *testing.T) {
	t.Parallel()

	rpc := &recordingProtocolClient{responses: map[string]any{
		"response.finished":         ResponseFinishedResult{},
		"response.security_details": ResponseSecurityDetailsResult{},
		"response.server_addr":      ResponseServerAddrResult{},
	}}
	response := newResponse(rpc, responseDescriptor())
	ctx := context.Background()

	if err := response.Finished(ctx); err != nil {
		t.Fatalf("Finished() error = %v", err)
	}
	if value, err := response.SecurityDetails(ctx); err != nil || value != nil {
		t.Fatalf("SecurityDetails() = (%#v, %v)", value, err)
	}
	if value, err := response.ServerAddr(ctx); err != nil || value != nil {
		t.Fatalf("ServerAddr() = (%#v, %v)", value, err)
	}
}
